book-content/chapters/14-configuring-typescript.md

```diff
@@ -0,0 +1,1024 @@
+  year: number;
+}
+
+export declare function createAlbum(
+  title: string,
+  artist: string,
+  year: number,
+): Album;
+```
+
+And the `album.js` file transpiled from TypeScript:
+
+```javascript
+// album.js
+"use strict";
+Object.defineProperty(exports, "__esModule", { value: true });
+exports.createAlbum = void 0;
+function createAlbum(title, artist, year) {
+  return { title, artist, year };
+}
+exports.createAlbum = createAlbum;
+```
+
+Now, anyone who uses your library will have access to the type information in the `.d.ts` files.
+
+### Declaration Maps
+
+One common use case for libraries is inside monorepos. A monorepo is a collection of libraries, each with their own `package.json`, that can be shared across different applications. This means that you'll often be developing the library and the application that uses it side-by-side.
+
+For example, a monorepo might look like this:
+
+```
+monorepo
+  ├── apps
+  │   └── my-app
+  │       └── package.json
+  └── packages
+      └── my-library
+          └── package.json
+```
+
+However, if you're working on code inside `my-app` that imports from `my-library`, you'll be working with the compiled JavaScript code, not the TypeScript source code. This means that when you `CMD + click` on an import, you'll be taken to the `.d.ts` file instead of the original TypeScript source.
+
+This is where declaration maps come in. They provide a mapping between the generated `.d.ts` and the original `.ts` source files.
+
+In order to create them, the `declarationMap` setting should be added to your `tsconfig.json` file, as well as the `sourceMap` setting:
+
+```json
+{
+  "compilerOptions": {
+    "declarationMap": true,
+    "sourceMap": true
+  }
+}
+```
+
+With this option in place, the TypeScript compiler will generate `.d.ts.map` files alongside the `.d.ts` files. Now, when you `CMD + click` on an import in `my-app`, you'll be taken to the original TypeScript source file in `my-library`.
+
+This is less useful when your library is on `npm`, unless you also ship your source files - but that's a little outside the scope of this book. In a monorepo, however, declaration maps are a great quality-of-life improvement.
+
+## `jsx`
+
+TypeScript has built-in support for transpiling JSX syntax, which is a syntax extension for JavaScript that allows you to write HTML-like code in your JavaScript files. In TypeScript, these need to be written in files with a `.tsx` extension:
+
+```tsx
+// Component.tsx
+const Component = () => {
+  return <div />;
+};
+```
+
+The `jsx` option tells TypeScript how to handle JSX syntax, and has five possible values. The most common are `preserve`, `react`, and `react-jsx`. Here's what each of them does:
+
+- `preserve`: Keeps JSX syntax as-is.
+- `react`: Transforms JSX into `React.createElement` calls. Useful for React 16 and earlier.
+- `react-jsx`: Transforms JSX into `_jsx` calls, and automatically imports from `react/jsx-runtime`. Useful for React 17 and later.
```

Comment:
which can be paired with `jsxImportSource` if one is using JSX with some other libraries like SolidJS